Addressing compile & runtime issues for plugin-style assemblies

ajaxx opened the pull request:
# Overview

This hotfix is designed to address two deficiencies identified while using dynamically loaded assemblies (e.g., plugins)

* Addresses inability to load references from hidden libraries.
* Addresses inability to load assemblies unknown at runtime.

## MetadataReferenceProvider

The first issue is that the compiler is unaware of references for types you may want to use but are not visible (for any number of reasons).  To address this we've added a `MetadataReferenceProvider`.  It allows the program to provide additional MetadataReferences (required for Roslyn at runtime).

To provide your own, add a reference to the container.  Reference code:

```
        container.Register<MetadataReferenceProvider>(
            () => myMetadataReferences, Lifespan.Singleton);
```

This allows your code to provide a delegate that is called to get your MetadataReferences at compile time. 

## AssemblyResolver

The second issue is that at runtime, NEsper may reference libraries that were compiled using the aforementioned MetadataReferences.  These references may fail making the runtime fail.  .NET provides a mechanism for handling this associated with the `AssemblyLoadContext`, but `AssemblyLoadContext` initialization is deep in the `ArtifactRepository`.  You can use the `AssemblyResolver` (like the `TypeResolver`) to resolve an assembly at runtime if the `ArtifactRepository` fails to resolve an assembly.

```
container.Register<AssemblyResolver>(ic => myResolveAssembly, Lifespan.Singleton);
```

# Example

I modified the SmartTypeResolver I previously provided as follows:

```
public class SmartTypeResolver : TypeResolver, TypeRepository
{
    private readonly TypeResolver _parentTypeResolver;
    private readonly IDictionary<string, Type> _typeCache;
    private readonly IDictionary<string, Assembly> _assemblyCache;

    public SmartTypeResolver()
    {
        // use the default resolver for resolving types if the type
        // is not known to this resolver
        _parentTypeResolver = TypeResolverDefault.INSTANCE;
        // a cache that contains types that we are aware of but may not
        // be visible to the default type resolver
        _typeCache = new Dictionary<string, Type>();

        _assemblyCache = new Dictionary<string, Assembly>();
    }

    public IEnumerable<MetadataReference> MetadataReferences {
        get {
            return _typeCache.Values
                .Select(_ => _.Assembly)
                .Distinct()
                .Select(_ => MetadataReference.CreateFromFile(_.Location))
                .Distinct()
                .ToList();
        }
    }

    public void AddType(Type type)
    {
        _typeCache.TryAdd(type.FullName!, type);
        _assemblyCache.TryAdd(type.Assembly.GetName().Name, type.Assembly);
    }

    public Assembly ResolveAssembly(AssemblyName assemblyName)
    {
        _assemblyCache.TryGetValue(assemblyName.Name, out var assembly);
        return assembly;
    }
    
    public Type ResolveType(
        string typeName,
        bool resolve = false)
    {
        if (_typeCache.TryGetValue(typeName, out var typeInstance))
        {
            return typeInstance;
        }

        var resolveType = resolve ? _parentTypeResolver.ResolveType(typeName, resolve) : null;
        return resolveType;
    }
}
```

And then I added the MetadataReferences and AssemblyResolver to the container.

```
container.Register<AssemblyResolver>(ic => smartTypeResolver.ResolveAssembly, Lifespan.Singleton);
container.Register<MetadataReferenceProvider>(
            () => smartTypeResolver.MetadataReferences, Lifespan.Singleton);
```

This allows the test application (with a plugin) to resolve external types at compile time and runtime.

ajaxx committed:
Addresses inability to load references from hidden libraries.
Addresses inability to load assemblies unknown at runtime.

TimeWanderer commented:
Thanks for the update, I will test it out ASAP.

TimeWanderer commented:
What is the purpose of resolve for ResolveType method? This is why I'm asking:
When I register a virtual window forge, it cannot resolve the class for that forge. Forge is supposed to be resolved in PluggableObjectCollection.HandleAddPluggableObject method. On line 206, the following code is called: 

`clazz = importService.TypeResolver.ResolveType(factoryClassName, false);`

Since resolve=false, SmartTypeResolver.ResolveType does not attempt to use parent resolver to resolve the type.

I'm sorry that I couldn't send a sample for this but it seemed like a fairly straightforward issue.

Edit: When I change the method call to use true instead of false, this part works correctly.

TimeWanderer commented:
We've come across another issue with StandardReaderWriterLock which seems to be stemming from the changes in [this commit](https://github.com/espertechinc/nesper/commit/66af3dda6e2fb5b29f38af70145d6e281c51c942) . At some point in deploying our modules, we get lock timeout. After we reverted these changes, it has been resolved.

It seems to be working functionally. But for some reason, it consumes insane amounts of memory just by compiling/deploying things. It's absurdly higher than earlier versions. We get 10+ gb of memory just from all the compilations and deployments. It was usually less than a gb for earlier versions. I am still investigating this issue. Is there anything you can think of that might be related to it?

TimeWanderer commented:
![image](https://github.com/espertechinc/nesper/assets/4564471/51003c49-1383-418c-a25d-0152d4aad917)

This is the result of profiling. There is no memory leak, but an immense GC pressure, which is all caused by unmanaged memory.

My first thought was too many calls to MetadataReference.CreateFromFile would create too many new objects each time it is called. So I cached the metadata references on AddType calls to a dictionary. It did not change the result. Since it's only unmanaged memory, I cannot analyze the call site or any other profiling data.

ajaxx commented:
> My first thought was too many calls to MetadataReference.CreateFromFile would create too many new objects each time it is called. So I cached the metadata references on AddType calls to a dictionary. It did not change the result. Since it's only unmanaged memory, I cannot analyze the call site or any other profiling data.

MetadataReferences are only called during compilation.  And that would only occur if you were continuously deploying new EPL statement.  So I doubt it's that.  I have a test harness that creates and deploys a few statements and then sends events.  The intent is to isolate event creation from statement creation.  And in that model I see no memory growth, in fact under dotmemory it's completely flat.  See the following.

![image](https://github.com/espertechinc/nesper/assets/827679/b7ffd047-c68d-496a-9343-f1d0a41ad504)

Can you help me build a harness that more accurately represents what you're doing throughout the entire lifecycle.

ajaxx commented:
> What is the purpose of resolve for ResolveType method? This is why I'm asking: When I register a virtual window forge, it cannot resolve the class for that forge. Forge is supposed to be resolved in PluggableObjectCollection.HandleAddPluggableObject method. On line 206, the following code is called:
> 
> `clazz = importService.TypeResolver.ResolveType(factoryClassName, false);`
> 
> Since resolve=false, SmartTypeResolver.ResolveType does not attempt to use parent resolver to resolve the type.
> 
> I'm sorry that I couldn't send a sample for this but it seemed like a fairly straightforward issue.
> 
> Edit: When I change the method call to use true instead of false, this part works correctly.

In most cases, the biggest difference is whether the TypeLoader attempts to use the parent for resolution.  There are some minor differences in behavior when it's set to true, in particular there's an assumption about materialization of a type that may not be loaded into space.  However, for this case, I generally agree with you that there's no reason why that flag is set to false.

I'd like to understand this one a little more because it seems as if materialization and recursion are being conjoined.  The TypeResolver should always recurse if it has a parent resolver.  But if the type does not exist due to being non-materialized then the flag should have kicked in.

I'll look at this one further.

TimeWanderer commented:
> > My first thought was too many calls to MetadataReference.CreateFromFile would create too many new objects each time it is called. So I cached the metadata references on AddType calls to a dictionary. It did not change the result. Since it's only unmanaged memory, I cannot analyze the call site or any other profiling data.
> 
> MetadataReferences are only called during compilation. And that would only occur if you were continuously deploying new EPL statement. So I doubt it's that. I have a test harness that creates and deploys a few statements and then sends events. The intent is to isolate event creation from statement creation. And in that model I see no memory growth, in fact under dotmemory it's completely flat. See the following.
> 
> ![image](https://user-images.githubusercontent.com/827679/254583149-b7ffd047-c68d-496a-9343-f1d0a41ad504.png)
> 
> Can you help me build a harness that more accurately represents what you're doing throughout the entire lifecycle.

I will try to prepare a sample by Friday to pinpoint the issue. I need to see what type of compilation, or less likely deployment, causes it in our context. The issue I see is that it allocates too much unmamanged memory to the point that GC cannot keep up for a long time. It is re-collected, given time.

Edit: I need to be more clear. The memory stays flat after the deployment is in place. But the action of compilation itself, or deployment, allocates too much memory. After it is deployed, the memory is flat.

ajaxx commented:
> > What is the purpose of resolve for ResolveType method? This is why I'm asking: When I register a virtual window forge, it cannot resolve the class for that forge. Forge is supposed to be resolved in PluggableObjectCollection.HandleAddPluggableObject method. On line 206, the following code is called:
> > `clazz = importService.TypeResolver.ResolveType(factoryClassName, false);`
> > Since resolve=false, SmartTypeResolver.ResolveType does not attempt to use parent resolver to resolve the type.
> > I'm sorry that I couldn't send a sample for this but it seemed like a fairly straightforward issue.
> > Edit: When I change the method call to use true instead of false, this part works correctly.
> 
> In most cases, the biggest difference is whether the TypeLoader attempts to use the parent for resolution. There are some minor differences in behavior when it's set to true, in particular there's an assumption about materialization of a type that may not be loaded into space. However, for this case, I generally agree with you that there's no reason why that flag is set to false.
> 
> I'd like to understand this one a little more because it seems as if materialization and recursion are being conjoined. The TypeResolver should always recurse if it has a parent resolver. But if the type does not exist due to being non-materialized then the flag should have kicked in.
> 
> I'll look at this one further.

So I looked into this one a bit further and my guess is that you were using the SmartTypeResolver code I provided earlier which has a bug.

```
    public Type ResolveType(
        string typeName,
        bool resolve = false)
    {
        if (_typeCache.TryGetValue(typeName, out var typeInstance))
        {
            return typeInstance;
        }

        var resolveType = _parentTypeResolver.ResolveType(typeName, resolve);
        return resolveType;
    }
```

And with this the resolution is passed to the parentTypeResolver with the resolve flag.  My mistake was I was short-circuiting the behavior based upon the resolve flag.

ajaxx commented:
> > > My first thought was too many calls to MetadataReference.CreateFromFile would create too many new objects each time it is called. So I cached the metadata references on AddType calls to a dictionary. It did not change the result. Since it's only unmanaged memory, I cannot analyze the call site or any other profiling data.
> > 
> > 
> > MetadataReferences are only called during compilation. And that would only occur if you were continuously deploying new EPL statement. So I doubt it's that. I have a test harness that creates and deploys a few statements and then sends events. The intent is to isolate event creation from statement creation. And in that model I see no memory growth, in fact under dotmemory it's completely flat. See the following.
> > ![image](https://user-images.githubusercontent.com/827679/254583149-b7ffd047-c68d-496a-9343-f1d0a41ad504.png)
> > Can you help me build a harness that more accurately represents what you're doing throughout the entire lifecycle.
> 
> I will try to prepare a sample by Friday to pinpoint the issue. I need to see what type of compilation, or less likely deployment, causes it in our context. The issue I see is that it allocates too much unmamanged memory to the point that GC cannot keep up for a long time. It is re-collected, given time.
> 
> Edit: I need to be more clear. The memory stays flat after the deployment is in place. But the action of compilation itself, or deployment, allocates too much memory. After it is deployed, the memory is flat.

If you can give me an idea of what you're doing I can also try to create a similar sample.  For example, are you creating a lot of objects that are injected into the engine.  If so, what's the rough frequency.

TimeWanderer commented:
Changing the smart type resolver fixed the issue of virtual window forge type resolution. 

For the issue of memory; I am creating around 100+ statements, most of which are deploying windows. Many of the windows are virtual. The rest are "create schema" and select statements.

TimeWanderer commented:
For example, compiling this module increased the memory by around 800 MB.

`@name('KeyboardInputControl') select * from pattern [every timer:interval(3 sec)];@name('HeartbeatEvent') select * from pattern [every timer:at(*, *, *, *, *, */1)];@name('EndOfDayEvent') select * from pattern [every timer:at(00, 17, *, *, *, 0, 'Eastern Standard Time' )];@name('StopEventHandler') select * from StopEvent;@name('StopApplication') select * from StopManagerEvent;@name('EspTerminationCommand') select * from ImmediateTerminationEvent;`

While the following module took 500 MB:

`@name(ApplicationConfigEventEventType) create schema ApplicationConfigEvent as RXI.Core.Applications.ApplicationConfigEvent;@name('ApplicationsWindow') create window _Applications#unique(Path) as ApplicationConfigEvent;@name('InsertApplicationConfigEvent') on ApplicationConfigEvent ce merge _Applications cw where ce.Path=cw.Path when not matched then insert select * when matched and ce.IsDelete=true then delete  when matched and ce.IsDelete=false then update set cw.Name=ce.Name, cw.Type=ce.Type, cw.IsValid=ce.IsValid, cw.VersionInfo=ce.VersionInfo, cw.Role=ce.Role, cw.ReadOnly=ce.ReadOnly, cw.Location=ce.Location, cw.IsDelete=ce.IsDelete;`

This is a very rough observation I made by taking a memory diff used by the process before and after the compilation. These might be related to many things, maybe the number and size of assemblies the compiler needs to compile these statements. Deployments don't affect the memory, so we can focus on compilations.

Edit: the total size of all the DLLs in the output program is around 50 MB.

TimeWanderer commented:
I think I found the culprit: RoslynCompiler.GetCoreMetadataReferences. In the old version, it used to cache the metadata references. In the new version, it rereads everything in the default load context and creates them from scratch. This creates a lot of memory pressure. This is probably overlooked in a small program since the default load context contains such a small binary that it doesn't make a dent. However, these 50-100 mbs of DLLs add up to lots of memory pressure.

You can immediately observe this by adding many library references, loading assemblies and checking memory values before and after this code line in RoslynCompiler.CompileInternal method:

`var metadataReferences = Enumerable.Concat(GetCoreMetadataReferences(), MetadataReferences);`

You need to call ToArray or ToList on the resulting enumerable to materialize it. In the current code, it hides in CSharpCompilation.Create call because it is materialized there.

In the end, it really was about the metadata references.

ajaxx committed:
Adding MetadataReferenceResolver to allow for alternate ways to load
MetadataReferences from Assemblies.

ajaxx commented:
I've added MetadataReferenceResolver to allow for deviation from the default way in which we resolve MetadataReferences.  You can use the IContainer to register an instance of a MetadataReferenceResolver and we will use that instead (in the RoslynCompiler) to resolve a MetadataReference.  Here is an example of how I've build a cached version of the MetadataReferenceResolver.

```
using System.Reflection;

using com.espertech.esper.common.@internal.util;
using com.espertech.esper.compat;
using Common;

using Microsoft.CodeAnalysis;

namespace Issue211a;

public class MetadataReferenceCache
{
    private readonly IDictionary<string, MetadataReference> _metadataCache;

    public MetadataReferenceCache()
    {
        _metadataCache = new Dictionary<string, MetadataReference>();
    }

    /// <summary>
    /// Gets a metadata reference from an assembly.  This method provides no caching of the metadata
    /// reference.
    /// </summary>
    /// <param name="assembly"></param>
    /// <returns>a metadata reference for the requested assembly</returns>
    public MetadataReference GetMetadataReference(Assembly assembly)
    {
        if (string.IsNullOrEmpty(assembly.Location)) {
            return null;
        }
        
        if (_metadataCache.TryGetValue(assembly.Location, out var metadataReference)) {
            return metadataReference;
        }

        return (_metadataCache[assembly.Location] = MetadataReference.CreateFromFile(assembly.Location));
    }
}
```

Using this is straightforward.  Simply register it with the Container.

```
        container.Register<MetadataReferenceResolver>(_metadataReferenceCache.GetMetadataReference, Lifespan.Singleton);
```

It's a delegate, so you can use a class or a lambda.

You might be asking why not cache everything.  There is a cost to generically caching and it would require a mechanism to clearly remove assemblies that have been unloaded.  This would require a lot of additional monitoring of the AssemblyLoadContext.  Reference management is generally handled by the runtime and so it's a lot of overhead for use cases that perform a significant amount of compilation.  However, the cache implementation allow you to control when you wish to evict the cached data.  In my opinion, this works much better.

TimeWanderer commented:
I this method called for nesper generated assemblies too if one nesper module uses types created by another nesper module(like inlined classes, etc.)? Should I cache those too or just the assemblies I load outside of nesper?

ajaxx committed:
Adding ability to provide your own core assemblies.

TimeWanderer commented:
Hello @ajaxx ,

I've managed to try it out. The memory fluctuation is not a problem anymore. There are ~~2~~ 3 issues now:
- There is a console log remaining in NamedWindowDispatchServiceImpl.Dispatch(), on line 98.
- There is significant, constant CPU usage. It feels like there is an idle loop on some thread because I can see one or two cores on Resource Monitor idling around 90% without a break. I am investigating it further. I will see if I can provide a sample.

Edit:
- NETCORE constant is not defined for the project if it's used by net6.0+ project, I had to add it manually. I think in this case, it is restored using net6.0, which does not add a NETCORE constant.

ajaxx committed:
Removing debugging console log statements.

ajaxx committed:
Adding net7.0 to Directory.Build.props

ajaxx commented:
> There is a console log remaining in NamedWindowDispatchServiceImpl.Dispatch(), on line 98.

I found a few other places where I had some lingering trace logging.  I've removed all of those.

> NETCORE constant is not defined for the project if it's used by net6.0+ project, I had to add it manually. I think in this case, it is restored using net6.0, which does not add a NETCORE constant.

Directory.Build.props contains the rules for defining the constant.  It appears I had not added net7.0 to the targetframeworks, but I've added it.  Currently, it appears that Microsoft has only defined net5.0, net6.0 and net7.0 - but let me know if you find a missing framework.

> There is significant, constant CPU usage. It feels like there is an idle loop on some thread because I can see one or two cores on Resource Monitor idling around 90% without a break. I am investigating it further. I will see if I can provide a sample.

This sounds like one of the timers.  The system timers are low CPU usage but they are also inaccurate and subject to drift.  Normally when I see high CPU utilization it's either the use of a timer that uses CPU to gain a more accurate clock cycle or it can be one of the spin-based queue processors.  Roughly how many events are you pushing through your system?  You can see this occurring during idle times (no or very low events pushing through)?

Please let me know if you're using any customizations and I'll try to do some similar testing.

TimeWanderer commented:
I've found the source. It's HarmonicTimer which is used in place of HighResolutionTimer. This is the code piece:
```
#if NETCORE
                    _harmonic = new HarmonicTimer(OnTimerEvent);
#else
                    _harmonic = new HighResolutionTimer(OnTimerEvent, null, 0, 10);
#endif
```

TimeWanderer commented:
When I removed it and used HightResolutionTimer, the CPU usage went back to normal.

ajaxx commented:
I assume the code to select the type of timer is in your code as I do not see it anywhere in the NEsper codebase.  This has been a long standing issue and I'll explain it briefly.  NEsper relies on metronomic behavior from timers.  If a timer is set to fire every 100 ms it should fire 10 times per second.  In Java, the behavior is metronomic, but in C# none of the timers are metronomic.  When NEsper was first written this was a big problem and we had to work around this by creating timer cycles that were smaller than the requested time-slice and then putting burst sleep or yields in to account for clock drift.

A little later, we were able to incorporate the windows multimedia timers.  These are a limited resource but for time-sensitive behaviors this was more than accurate.

Unfortunately, most systems are not nearly as time-sensitive as these use cases demonstrate.  As such, a standard system timer will work well for a great number of use cases.  In fact, if you are not sensitive to occasionally seeing some drift (e.g., 5 calls per second versus 6) then you can just use a system timer and your CPU usage will be almost nothing.

TimeWanderer commented:
[This is the line](https://github.com/espertechinc/nesper/blob/f72bd0ec39fccaea2b21ec5f05b2c57370307464/src/NEsper.Compat/compat/timers/SystemTimerFactory.cs#L104) I was talking about.  We do not use this factory directly, but we use crontab-based patterns via nesper.

ajaxx committed:
Fixing the precompilation directives.
Setting the default timer factory.

ajaxx commented:
I've changed the default behavior of the SystemTimerFactory and I've fixed most of the precompilation directives.  This PR should be ready for rebase and merge.

TimeWanderer commented:
Hello Ajaxx,

Thanks for all the hardwork. These are the remaining problems that I've found, some of which I've created support tickets earlier on. First point is related to the recent update, others might be overlooked in the latest updates.

- IsDebugOptimization in compiler configuration has private setter.
- NamedWindowTailViewInstance SnapshotDelete, SnapshotUpdate: These should use write locks because it's a write operation.
- VirtualDWEventTable.GetEnumerator is returning empty result.
- In FAFQueryMethodSelectExecNoContextJoin.Execute, ProcessJoinResult is called with null value. In some cases this causes null exception in generated ResultSetProcessor implementation.

Best regards

ajaxx commented:
Thank you for the heads up on the remaining items.

> IsDebugOptimization in compiler configuration has private setter.

I've added a public setter.  Also, it wasn't being set when the RoslynCompiler is created, so I've fixed that also.

> NamedWindowTailViewInstance SnapshotDelete, SnapshotUpdate: These should use write locks because it's a write operation.

I'll discuss this with Tom.  I need to understand the scope of the lock as it may only acquiring the lock on the data being read, no the data being modified.

> VirtualDWEventTable.GetEnumerator is returning empty result.

Expected behavior, but there was some dead code here that was not consistent with how Esper handles this.  I've fixed this.

> In FAFQueryMethodSelectExecNoContextJoin.Execute, ProcessJoinResult is called with null value. In some cases this causes null exception in generated ResultSetProcessor implementation.

I've changed this to use the `EmptySet<MultiKeyArrayOfKeys<EventBean>>.Instance` which will always provide an instance that has no data.

ajaxx committed:
Cleaning up some residual issues.
Making SerializerFactory singleton rather than purely static (upcoming changes).

TimeWanderer commented:
Hello @ajaxx,

Sorry to bother you with this further.  If `VirtualDWEventTable.GetEnumerator` is supposed to return an empty result, does it mean that a join snapshot with virtual windows is also supposed to return an empty result? 
I am asking this because we are joining virtual window data. When we actually return the enumerator of virtual window data from `VirtualDWQueryPlanUtil.GetEnumerator` via `VirtualDWViewMayNull.VirtualDataWindow.GetEnumerator()`, it produces the result we want.

Thanks for the update.
